session.go

```diff
@@ -682,8 +685,11 @@ func (s *Session) allocateHandle(l *link) error {
 
 	next, ok := s.handles.Next()
 	if !ok {
+		if err := s.Close(ctx); err != nil {
+			return err
+		}
 		// handle numbers are zero-based, report the actual count
-		return fmt.Errorf("reached session handle max (%d)", s.handleMax+1)
+		return &SessionError{inner: fmt.Errorf("reached session handle max (%d)", s.handleMax+1)}
```

Comment:
For callers (ie, me) I should treat this as requiring a connection reset - right? Would it make more sense to return this as a ConnError? This would allow us to reserve SessionError for things that only affect the individual session.

Reply:
Handle are per session, so you should only need to recreate the session.

Reply:
btw, s.handleMax+1 will just rollover and be zero, right? Do we want to the +1?

Reply:
You are correct.  However, this has made me realize there's a screwy corner case here.

`SessionOptions.MaxLinks` is the maximum number of links for a session.  The link handles are zero-based, so if you specify 10, then the max link handle is 9.  We calculate the offset in `newSession()`.

The default value is `math.MaxUint32`, but due to the above offset calculation, if you were to specify `math.MaxUint32`, then you'd get a zero-based number of max handles and the `-1` in the debug log would work as expected.

I think we should change the default value to be `math.MaxUnit32-1` so that the behavior is consistent and there's no rollover in the log for the default value.  What do you think?

Reply:
Sounds good to me.